pkg/analytics/instrumentation_collector.go

```diff
@@ -0,0 +1,242 @@
+	ic.interactionId = id
+}
+
+func (ic *instrumentationCollectorImpl) SetTimestamp(t time.Time) {
+	ic.timestamp = t
+}
+
+func (ic *instrumentationCollectorImpl) SetDuration(d time.Duration) {
+	ic.duration = d
+}
+
+func (ic *instrumentationCollectorImpl) SetStage(s string) {
+	ic.stage = s
+}
+
+func (ic *instrumentationCollectorImpl) SetType(t string) {
+	ic.instrumentationType = t
+}
+
+func (ic *instrumentationCollectorImpl) SetCategory(c []string) {
+	ic.category = c
+}
+
+func (ic *instrumentationCollectorImpl) SetStatus(s Status) {
+	ic.status = s
+}
+
+func (ic *instrumentationCollectorImpl) SetTestSummary(s json_schemas.TestSummary) {
+	ic.testSummary = s
+}
+
+func (ic *instrumentationCollectorImpl) SetTargetId(t string) {
+	ic.targetId = t
+}
+
+func (ic *instrumentationCollectorImpl) AddError(err error) {
+	ic.instrumentationErr = append(ic.instrumentationErr, err)
+}
+
+func (ic *instrumentationCollectorImpl) AddExtension(key string, value interface{}) {
+	if ic.extension == nil {
+		ic.extension = make(map[string]interface{})
+	}
+	ic.extension[key] = value
+}
+
+func GetV2InstrumentationObject(collector InstrumentationCollector) (*api.AnalyticsRequestBody, error) {
+	t, ok := collector.(*instrumentationCollectorImpl)
+	if ok {
+		return t.GetV2InstrumentationObject()
+	}
+	return nil, fmt.Errorf("failed to convert collector")
+}
+
+func (ic *instrumentationCollectorImpl) GetV2InstrumentationObject() (*api.AnalyticsRequestBody, error) {
+	a, err := ic.getV2Attributes()
+
+	if err != nil {
+		return nil, err
+	}
+
+	d := api.AnalyticsData{
+		Type:       ic.instrumentationType,
+		Attributes: a,
+	}
+
+	return &api.AnalyticsRequestBody{
+		Data: d,
+	}, nil
+}
+
+func (ic *instrumentationCollectorImpl) getV2Attributes() (api.AnalyticsAttributes, error) {
+	r, err := ic.getV2Runtime()
+
+	if err != nil {
+		return api.AnalyticsAttributes{}, err
+	}
+
+	return api.AnalyticsAttributes{
+		Interaction: ic.getV2Interaction(),
+		Runtime:     r,
+	}, nil
+}
+
+func (ic *instrumentationCollectorImpl) getV2Interaction() api.Interaction {
+	return api.Interaction{
+		Categories:  &ic.category,
+		Errors:      toInteractionErrors(ic.instrumentationErr),
+		Extension:   &ic.extension,
+		Id:          ic.interactionId,
+		Results:     toInteractionResults(&ic.testSummary),
+		Stage:       toInteractionStage(&ic.stage),
+		Status:      string(ic.status),
+		Target:      api.Target{Id: ic.targetId},
+		TimestampMs: ic.timestamp.UnixMilli(),
+		Type:        ic.instrumentationType,
+	}
+}
+
+func (ic *instrumentationCollectorImpl) getV2Runtime() (*api.Runtime, error) {
+	hasUserAgentAppData := len(ic.userAgent.App) > 0 || len(ic.userAgent.AppVersion) > 0
+	if !hasUserAgentAppData {
+		return nil, errors.New("no user agent application data")
+	}
+
+	return &api.Runtime{
+		Application: &api.Application{
+			Name:    ic.userAgent.App,
+			Version: ic.userAgent.AppVersion,
+		},
+		Environment: &api.Environment{
```

Comment:
Issue: some of the data is optional and might not be available, so we should not add empty strings here.